### PR TITLE
[SPARK-43390][SQL] DSv2 allows CTAS/RTAS to reserve schema nullability

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

Add a new method `useNullableQuerySchema` in `TableCatalog`, to allow the DataSource implementation to declare whether they need to reserve schema nullability on CTAS/RTAS.

### Why are the changes needed?

SPARK-28837 forcibly uses the nullable schema on CTAS/RTAS, which seems too aggressive:

1. The existing matured RDBMSs have different behaviors for reserving schema nullability on CTAS/RTAS, as mentioned in #25536, PostgreSQL forcibly uses the nullable schema, but MySQL respects the query's output schema nullability.
2. Some OLAP systems(e.g. ClickHouse) are perf-sensitive for nullable, and have strict restrictions on table schema, e.g. the primary keys are not allowed to be nullable.

### Does this PR introduce _any_ user-facing change?

Yes, this PR adds a new DSv2 API, but the default implementation reserves backward compatibility.

### How was this patch tested?

UTs are updated.

### CODE REVIEW

**Comment by pan3793** [comment]
cc @cloud-fan @sunchao @beliefer

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java [hunk 1]
```diff
@@ -199,6 +199,14 @@ default Table createTable(
     return createTable(ident, CatalogV2Util.v2ColumnsToStructType(columns), partitions, properties);
   }
 
+  /**
+   * Return whether to reserve schema nullability of query output or forcibly use nullable schema
+   * on creating table implicitly, e.g. CTAS/RTAS.
+   */
+  default boolean createTableReserveSchemaNullability() {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java:206]: how about
```
/**
 * If true, mark all the fields of the query schema as nullable when executing
 * CREATE/REPLACE TABLE ... AS SELECT ... and creating the table.
 */
default boolean useNullableQuerySchema
```

Comment (pan3793) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java]: Good suggestion, changed

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 2]
```diff
@@ -3261,3 +3278,7 @@ class FakeV2Provider extends SimpleTableProvider {
     throw new UnsupportedOperationException("Unnecessary for DDL tests")
   }
 }
+
+class TestCreateTableReserveSchemaNullabilityCatalog extends InMemoryCatalog {
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3282]: How about `ForbidSchemaNullabilityCatalog` ?

Comment (pan3793) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3282]: "forbid" seems not accurate here, would prefer to use "reserve" to represent respecting the query schema nullability

Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3282]: I means the name seems too long. Could you rename it with simple one ?

Comment (pan3793) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3282]: Maybe "ReserveSchemaNullabilityCatalog"?

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Review by beliefer — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**